<a href="https://colab.research.google.com/github/Nargiz-chess/Solar_forecasting/blob/main/Base_models_Germany.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
#Mount google drive
from google.colab import drive
drive.mount('/content/drive')

from pandas import read_csv
import pandas as pd
import datetime as dt
import matplotlib
from sklearn.linear_model import LinearRegression
import numpy as np
from sklearn import metrics
import matplotlib.pyplot as plt


Mounted at /content/drive


In [ ]:
path = "/content/drive/MyDrive/Master_thesis/Datasets/Merged(GER+weather).csv"
df = read_csv(path, header =0, index_col=1) 

In [ ]:
len(df)
df = df.fillna(0)

#Linear Regression

In [ ]:
from sklearn.linear_model import LinearRegression
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import MinMaxScaler
from statistics import mean
from sklearn import metrics

#perform a robust scaler transform of the dataset
scaler = MinMaxScaler()

X = df["radiation_surface"]

# for all the houses
X_reshaped = X.values.reshape(-1,1)
RMSEs=list()

for col in df.columns:
  if "pv" in col:
    y = df[col].fillna(0)
    y_reshaped = y.values.reshape(-1,1)
    y_reshaped = scaler.fit_transform(y_reshaped)
    X_train, X_test, y_train, y_test = train_test_split( X_reshaped, y_reshaped, test_size=24, shuffle = False)
    X_train = scaler.fit_transform(X_train)
    X_test = scaler.fit_transform(X_test)
    y_train = scaler.fit_transform(y_train)
    y_test = scaler.fit_transform(y_test)
    reg = LinearRegression().fit(X_train, y_train)
    print(col)
    print("R^2 of Linear Regression on the train set =", reg.score(X_train, y_train))
    print("R^2 of Linear Regression on the test set =", reg.score(X_test, y_test))

    y_pred = reg.predict(X_test)
    print("MAE:", metrics.mean_absolute_error(y_test,y_pred))
    print("MSE:", metrics.mean_squared_error(y_test,y_pred))
    rmse = np.sqrt(metrics.mean_squared_error(y_test,y_pred))
    RMSEs.append(rmse)
    print("RMSE:", rmse)
    print()

print("Average RMSE = ", mean(RMSEs))

DE_KN_industrial1_pv_2
R^2 of Linear Regression on the train set = 0.7558641578043406
R^2 of Linear Regression on the test set = -0.5863744573153902
MAE: 0.15450931616445288
MSE: 0.0794770848998525
RMSE: 0.2819168049262982

DE_KN_industrial2_pv
R^2 of Linear Regression on the train set = 0.8015194135718199
R^2 of Linear Regression on the test set = 0.7252103650186734
MAE: 0.08685068335072632
MSE: 0.023062341493168825
RMSE: 0.1518629036110163

DE_KN_industrial3_pv_facade
R^2 of Linear Regression on the train set = 0.6554373004143315
R^2 of Linear Regression on the test set = 0.7534756550343273
MAE: 0.09845813067998432
MSE: 0.02509003171043286
RMSE: 0.1583983324105177

DE_KN_industrial3_pv_roof
R^2 of Linear Regression on the train set = 0.8296482514593192
R^2 of Linear Regression on the test set = 0.8264351870718564
MAE: 0.0808393427791849
MSE: 0.019549645363581128
RMSE: 0.13982004635809964

DE_KN_residential1_pv
R^2 of Linear Regression on the train set = 0.7930216850801259
R^2 of Line

#Gradient Boosting Regressor

In [ ]:
from sklearn.ensemble import GradientBoostingRegressor

#GBR = GradientBoostingRegressor(random_state=None, max_depth = 3,learning_rate=0.1).fit(X_train, y_train)

# print("R^2 of Gradient Boosting Regressor on the training set =",GBR.score(X_train, y_train))
# print("R^2 of Gradient Boosting Regressor on the test set =",GBR.score(X_test, y_test))


X = df["radiation_surface"]

# for all the houses
X_reshaped = X.values.reshape(-1,1)
RMSEs_gbr=list()
r_squared=[]

for col in df.columns:
  if "pv" in col:
    y = df[col].fillna(0)
#    y_reshaped = y.values.reshape(-1,1)
#    y_reshaped = scaler.fit_transform(y_reshaped)
    X_train, X_test, y_train, y_test = train_test_split( X_reshaped, y, test_size=24, shuffle = False)
    # X_train = scaler.fit_transform(X_train)
    # X_test = scaler.fit_transform(X_test)
    # y_train = scaler.fit_transform(y_train)
    # y_test = scaler.fit_transform(y_test)
    gbr = GradientBoostingRegressor(random_state=None, max_depth = 3,learning_rate=0.1).fit(X_train, y_train)
    # print(col)
    # print("R^2 of Linear Regression on the train set =", gbr.score(X_train, y_train))
    # print("R^2 of Linear Regression on the test set =", gbr.score(X_test, y_test))

    y_pred = gbr.predict(X_test)
    # print("MAE:", metrics.mean_absolute_error(y_test,y_pred))
    # print("MSE:", metrics.mean_squared_error(y_test,y_pred))
    r_2=gbr.score(X_train, y_train)
    r_squared.append(r_2)
    rmse = np.sqrt(metrics.mean_squared_error(y_test,y_pred))
    RMSEs_gbr.append(rmse)
    # print("RMSE:", rmse)
    # print()

print("Average RMSE accross all columns = ", round(mean(RMSEs_gbr),2))
print("Average R^2 accross all columns = ", round(mean(r_squared),2))

Average RMSE accross all columns =  0.13
Average R^2 accross all columns =  0.81


#Gradient boosting - Hyper parameter tuning

In [ ]:
from sklearn.model_selection import GridSearchCV

#params = [{'loss': ['ls','huber'], 'learning_rate': [0.05, 0.07, 0.15, 0.2], 'n_estimators': [200],'max_depth': [5], 'subsample' : [1] },{'loss': ['ls','huber'], 'learning_rate': [0.05, 0.07, 0.2], 'n_estimators': [350],'max_depth': [6], 'subsample' : [1] },{'loss': ['ls','huber'], 'n_estimators': [100],'learning_rate': [0.1], 'max_depth': [4], 'subsample' : [1] }]
params = {'loss': ['ls', 'lad','huber','quantile'],'learning_rate': [0.1, 0.2,0.3],'max_depth': [3,5,6]} 

#gbr = GradientBoostingRegressor(random_state=None)

# print("R^2 of Gradient Boosting Regressor on the training set =",GBR.score(X_train, y_train))
# print("R^2 of Gradient Boosting Regressor on the test set =",GBR.score(X_test, y_test))


X = df["radiation_surface"]
gbr_best_scores = []
grb_best_params = []

# for all the houses
X_reshaped = X.values.reshape(-1,1)
RMSEs_gbr=list()

for col in df.columns:
  if "pv" in col:
    y = df[col].fillna(0)
    X_train, X_test, y_train, y_test = train_test_split( X_reshaped, y, test_size=24, shuffle = False)
    gbr = GridSearchCV(GradientBoostingRegressor(random_state=None), params, scoring = 'explained_variance', cv = 5, n_jobs = -1)
    gbr.fit(X_train,y_train)
    gbr_best_scores.append([col, gbr.best_score_])
    grb_best_params.append([col, gbr.best_params_])
print("Best Scores : ", gbr_best_scores )
print("Best Parameters : ",grb_best_params)


Best Scores :  [['DE_KN_industrial1_pv_2', 0.5803883490286398], ['DE_KN_industrial2_pv', 0.6889587044115786], ['DE_KN_industrial3_pv_facade', 0.6032783360235119], ['DE_KN_industrial3_pv_roof', -0.032842451327424804], ['DE_KN_residential1_pv', 0.6911895984362616], ['DE_KN_residential3_pv', 0.5597570650330612], ['DE_KN_residential4_pv', 0.6907473739246542], ['DE_KN_residential6_pv', 0.6617529607876806]]
Best Parameters :  [['DE_KN_industrial1_pv_2', {'learning_rate': 0.1, 'loss': 'ls', 'max_depth': 3}], ['DE_KN_industrial2_pv', {'learning_rate': 0.2, 'loss': 'lad', 'max_depth': 3}], ['DE_KN_industrial3_pv_facade', {'learning_rate': 0.1, 'loss': 'ls', 'max_depth': 3}], ['DE_KN_industrial3_pv_roof', {'learning_rate': 0.1, 'loss': 'huber', 'max_depth': 3}], ['DE_KN_residential1_pv', {'learning_rate': 0.1, 'loss': 'lad', 'max_depth': 5}], ['DE_KN_residential3_pv', {'learning_rate': 0.1, 'loss': 'huber', 'max_depth': 3}], ['DE_KN_residential4_pv', {'learning_rate': 0.1, 'loss': 'lad', 'max_de

Best Scores :  [['DE_KN_industrial1_pv_2', 0.5803883490286398], ['DE_KN_industrial2_pv', 0.6889587044115786], ['DE_KN_industrial3_pv_facade', 0.6032783360235119], ['DE_KN_industrial3_pv_roof', -0.032842451327424804], ['DE_KN_residential1_pv', 0.6911895984362616], ['DE_KN_residential3_pv', 0.5597570650330612], ['DE_KN_residential4_pv', 0.6907473739246542], ['DE_KN_residential6_pv', 0.6617529607876806]]

Best Parameters :  [['DE_KN_industrial1_pv_2', {'learning_rate': 0.1, 'loss': 'ls', 'max_depth': 3}], ['DE_KN_industrial2_pv', {'learning_rate': 0.2, 'loss': 'lad', 'max_depth': 3}], ['DE_KN_industrial3_pv_facade', {'learning_rate': 0.1, 'loss': 'ls', 'max_depth': 3}], ['DE_KN_industrial3_pv_roof', {'learning_rate': 0.1, 'loss': 'huber', 'max_depth': 3}], ['DE_KN_residential1_pv', {'learning_rate': 0.1, 'loss': 'lad', 'max_depth': 5}], ['DE_KN_residential3_pv', {'learning_rate': 0.1, 'loss': 'huber', 'max_depth': 3}], ['DE_KN_residential4_pv', {'learning_rate': 0.1, 'loss': 'lad', 'max_depth': 3}], ['DE_KN_residential6_pv', {'learning_rate': 0.1, 'loss': 'ls', 'max_depth': 3}]]

#Quantile regression

In [ ]:
from statsmodels.regression.quantile_regression import QuantReg
import statsmodels.api as sm
from statistics import mean

X = df["radiation_surface"]
X_reshaped = X.values.reshape(-1,1)
#quantiles can be added
quantile = [0.9]
RMSEs_QR=[]

for col in df.columns:
  if "pv" in col:
    y = df[col].fillna(0)
    y_reshaped = y.values.reshape(-1,1)
    y_reshaped = scaler.fit_transform(y_reshaped)
    X_train, X_test, y_train, y_test = train_test_split( X_reshaped, y_reshaped, test_size=24, shuffle = False)
    for q in quantile:
      y_pred = sm.QuantReg(y_train, X_train).fit(q=q).predict(X_test)
      print("quantile = ",q)
      print("MAE:", metrics.mean_absolute_error(y_test,y_pred))
      print("MSE:", metrics.mean_squared_error(y_test,y_pred))
      rmse = np.sqrt(metrics.mean_squared_error(y_test,y_pred))
      RMSEs_QR.append([q,rmse])
      print("RMSE:", rmse)
      print()

print("Average RMSE", mean(RMSEs_QR[1]))

quantile =  0.9
MAE: 0.025475685041147417
MSE: 0.002183032345575505
RMSE: 0.04672293168857777

quantile =  0.9
MAE: 0.026078883146876505
MSE: 0.002287022801758641
RMSE: 0.0478228272037386

quantile =  0.9
MAE: 0.024812530917352863
MSE: 0.002033328108418596
RMSE: 0.045092439592670036

quantile =  0.9
MAE: 0.02066238634396912
MSE: 0.0014580344545063399
RMSE: 0.03818421734835402

quantile =  0.9
MAE: 0.0181157513039111
MSE: 0.0012197083124769025
RMSE: 0.03492432264879167

quantile =  0.9
MAE: 0.020523616655417087
MSE: 0.0014491932132450448
RMSE: 0.03806827042623614

quantile =  0.9
MAE: 0.025278744426784978
MSE: 0.0021580104170405815
RMSE: 0.046454390718645544

quantile =  0.9
MAE: 0.02975738061004703
MSE: 0.0029429494263876776
RMSE: 0.054248957836880864

Average RMSE 0.47391141360186934


#Ridge Regression

In [ ]:
from sklearn.linear_model import Ridge

X = df["radiation_surface"]
X_reshaped = X.values.reshape(-1,1) 

y = df['DE_KN_residential6_pv']fillna(0)
y_reshaped = y.values.reshape(-1,1)

StandardScaler()
ridge = Ridge(alpha=1, fit_intercept=True)
ridge.fit(X_train, y_train)

print("R^2 of Linear Regression on the train set =", ridge.score(X_train, y_train))
print("R^2 of Linear Regression on the test set =", ridge.score(X_test, y_test))

(6954, 1)